### PR TITLE
Hash dicts (rebased)

### PR DESCRIPTION
**Pull request by MatMoore**
This is a rebased version of https://github.com/openregister/objecthash-java/pull/2

I've left in the static methods as helpers but I think that we will need to instantiate classes when hashing collection objects (see the example in the test)

The hashing of dicts is defined in this RFC:
https://github.com/openregister/registers-rfcs/blob/master/content/item-hash/index.md

I've interpreted this as:

1. Hash the key and the value for every value
2. Concat the keyHash and valueHash of every pair
3. Sort the resulting list of concatted hashes
4. Prefix the list with "d"
5. Hash the result

In order to implement this I needed a way to define the type of dict values so I've refactored the library to use an interface `ObjectHashable` and classes for each objecthash type.

For the moment I'm ignoring null values, but I'm not doing anything special with empty strings. I'm thinking that our treatment of empty strings in registers is part of [item normalisation](https://github.com/openregister/registers-rfcs/blob/master/content/blob-normalisation/index.md) and it should be left out of the objecthash code.

### CODE REVIEW

File: src/test/java/uk/gov/objecthash/ObjectHashTest.java [hunk 1]
```diff
@@ -35,4 +38,14 @@ public void hashesTimestamps() {
         Instant instant = date.atStartOfDay(ZoneId.of("UTC")).toInstant();
         assertEquals(ObjectHash.toHexDigest(instant), "cb34961d3d1a44386a73c37eb64c72a2b61ff40fc108abca92ef07c4954a1645" );
     }
+    
+    @Test
+    public void hashesDict() throws NoSuchAlgorithmException {
```
Comment (gidsg) [inline src/test/java/uk/gov/objecthash/ObjectHashTest.java]: this `throws` can be deleted

File: src/main/java/uk/gov/objecthash/DictValue.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import java.security.MessageDigest;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DictValue implements ObjectHashable {
+    private final Map<String, ObjectHashable> data;
+
+    public DictValue() {
```
Comment (gidsg) [inline src/main/java/uk/gov/objecthash/DictValue.java]: seems like this is unused?

File: src/main/java/uk/gov/objecthash/TimestampValue.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package uk.gov.objecthash;
+
+import java.time.Instant;
+
+public class TimestampValue implements ObjectHashable {
+
+    private final Instant value;
+
+    public TimestampValue(Instant value) {
```
Comment (gidsg) [inline src/main/java/uk/gov/objecthash/TimestampValue.java:9]: this `public` modifier, I think can be removed?

File: src/main/java/uk/gov/objecthash/DictValue.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        
+        List<byte[]> hashValues = new ArrayList<>();
+        for (Map.Entry<String, ObjectHashable> entry : this.data.entrySet()) {
+            ObjectHashable value = entry.getValue();
+            ObjectHashable key = new StringValue(entry.getKey());
+            if (value != null) {
+                hashValues.add(concatEntryHashes(key.digest(), value.digest()));
+            }
+        }
+        hashValues.sort((a, b) -> compareBytes(a, b));
```
Comment (gidsg) [inline src/main/java/uk/gov/objecthash/DictValue.java]: can be simplified to `(this::compareBytes)`

File: src/main/java/uk/gov/objecthash/DictValue.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        sha256.update(ObjectHashable.DICT_TAG.getBytes(StandardCharsets.UTF_8));
+        hashValues.stream().forEach(bytes -> sha256.update(bytes));
```
Comment (gidsg) [inline src/main/java/uk/gov/objecthash/DictValue.java]: Can be simplified to `(sha256::update)`

Comment (gidsg) [reply src/main/java/uk/gov/objecthash/DictValue.java]: I don't think `stream()` is needed here, i.e. can be simplified to `hashValues.forEach`

File: src/main/java/uk/gov/objecthash/TimestampValue.java (hunk 3 above)
Comment (MatMoore) [reply src/main/java/uk/gov/objecthash/TimestampValue.java:9]: will this be accessible outside of the package? Atm in order to hash a Map you need to build it up from primitives

eg

```java
        Map<String, ObjectHashable> data = new HashMap<>();
        data.put("foo", new StringValue("bar"));
        data.put("null", null);
```

Comment (gidsg) [reply src/main/java/uk/gov/objecthash/TimestampValue.java:9]: Ah, OK I wasn't sure if this was intended to be part of the public API?

**Review by gidsg — APPROVED** [review]
:+1:

**Pull request merged**